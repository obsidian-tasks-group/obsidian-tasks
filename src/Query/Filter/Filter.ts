import type { Task } from '../../Task';

/**
 * A filtering function, that takes a Task object and returns
 * whether it matches a particular filtering instruction.
 */
export type Filter = (task: Task) => boolean;

/**
 * A class which stores one of:
 * - A Filter
 * - An error message
 *
 * This is really currently a convenience for returning date from
 * Field.createFilterOrErrorMessage() and derived classes.
 *
 * Later, it may gain helper functions for constructing parser error messages,
 * as currently these are created by some rather repetitious code, and also
 * there is scope for making these messages more informative (including the
 * problem line, and perhaps listing allowed options).
 */
export class FilterOrErrorMessage {
    filter: Filter | undefined;
    error: string | undefined;

    /**
     * Construct a FilterOrErrorMessage with the filter.
     * @param filter
     */
    public static fromFilter(filter: Filter): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.filter = filter;
        return result;
    }

    /**
     * Construct a FilterOrErrorMessage with the given error message.
     * @param errorMessage
     */
    public static fromError(errorMessage: string): FilterOrErrorMessage {
        const result = new FilterOrErrorMessage();
        result.error = errorMessage;
        return result;
    }
}
