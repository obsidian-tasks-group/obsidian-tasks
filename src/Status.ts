import { StatusConfiguration, StatusType } from './StatusConfiguration';
import type { StatusCollectionEntry } from './StatusCollection';

/**
 * Tracks the possible states that a task can be in.
 *
 * Related classes:
 * @see StatusConfiguration
 * @see StatusRegistry
 * @see StatusSettings
 * @see StatusSettingsHelpers.ts
 * @see CustomStatusModal
 *
 * @export
 * @class Status
 */
export class Status {
    /**
     * The default Done status. Goes to Todo when toggled.
     *
     * @static
     * @type {Status}
     * @memberof Status
     */
    public static DONE: Status = Status.makeDone();

    /**
     * A default status of empty, used when things go wrong.
     *
     * @static
     * @memberof Status
     */
    public static EMPTY: Status = Status.makeEmpty();

    /**
     * The default Todo status. Goes to Done when toggled.
     * User may later be able to override this to go to In Progress instead.
     *
     * @static
     * @type {Status}
     * @memberof Status
     */
    public static TODO: Status = Status.makeTodo();

    /**
     * The configuration stored in the data.json file.
     *
     * @type {StatusConfiguration}
     * @memberof Status
     */
    public readonly configuration: StatusConfiguration;

    /**
     * The symbol used between the two square brackets in the markdown task.
     *
     * @type {string}
     * @memberof Status
     */
    public get symbol(): string {
        return this.configuration.symbol;
    }

    /**
     * Returns the name of the status for display purposes.
     *
     * @type {string}
     * @memberof Status
     */
    public get name(): string {
        return this.configuration.name;
    }

    /**
     * Returns the next status for a task when toggled.
     *
     * @type {string}
     * @memberof Status
     */
    public get nextStatusSymbol(): string {
        return this.configuration.nextStatusSymbol;
    }

    /**
     * If true then it is registered as a command that the user can map to.
     *
     * @type {boolean}
     * @memberof Status
     */
    public get availableAsCommand(): boolean {
        return this.configuration.availableAsCommand;
    }

    /**
     * Returns the status type. See {@link StatusType} for details.
     */
    public get type(): StatusType {
        return this.configuration.type;
    }

    /**
     * Creates an instance of Status. The registry will be added later in the case
     * of the default statuses.
     *
     * @param {StatusConfiguration} configuration
     * @memberof Status
     */
    constructor(configuration: StatusConfiguration) {
        this.configuration = configuration;
    }

    /**
     * The default Done status. Goes to Todo when toggled.
     */
    static makeDone(): Status {
        return new Status(new StatusConfiguration('x', 'Done', ' ', true, StatusType.DONE));
    }

    /**
     * A default status of empty, used when things go wrong.
     */
    static makeEmpty(): Status {
        return new Status(new StatusConfiguration('', 'EMPTY', '', true, StatusType.EMPTY));
    }

    /**
     * The default Todo status. Goes to Done when toggled.
     * User may later be able to override this to go to In Progress instead.
     */
    static makeTodo(): Status {
        return new Status(new StatusConfiguration(' ', 'Todo', 'x', true, StatusType.TODO));
    }

    /**
     * The default Cancelled status. Goes to Todo when toggled.
     */
    static makeCancelled(): Status {
        return new Status(new StatusConfiguration('-', 'Cancelled', ' ', true, StatusType.CANCELLED));
    }

    /**
     * The default In Progress status. Goes to Done when toggled.
     */
    static makeInProgress(): Status {
        return new Status(new StatusConfiguration('/', 'In Progress', 'x', true, StatusType.IN_PROGRESS));
    }

    /**
     * Return the StatusType to use for a symbol, if it is not in the StatusRegistry.
     * The core symbols are recognised.
     * Other symbols are treated as StatusType.TODO
     * @param symbol
     */
    static getTypeForUnknownSymbol(symbol: string): StatusType {
        switch (symbol) {
            case 'x':
            case 'X':
                return StatusType.DONE;
            case '/':
                return StatusType.IN_PROGRESS;
            case '-':
                return StatusType.CANCELLED;
            case '':
                return StatusType.EMPTY;
            case ' ':
            default:
                return StatusType.TODO;
        }
    }

    /**
     * Convert text that was saved from a StatusType value back to a StatusType.
     * Returns StatusType.TODO if the string is not valid.
     * @param statusTypeAsString
     */
    static getTypeFromStatusTypeString(statusTypeAsString: string): StatusType {
        return StatusType[statusTypeAsString as keyof typeof StatusType] || StatusType.TODO;
    }

    /**
     * Create a Status representing the given, unknown symbol.
     *
     * This can be useful when StatusRegistry does not recognise a symbol,
     * and we do not want to expose the user's data to the Status.EMPTY status.
     *
     * The type is set to TODO.
     * @param unknownSymbol
     */
    static createUnknownStatus(unknownSymbol: string) {
        return new Status(new StatusConfiguration(unknownSymbol, 'Unknown', 'x', false, StatusType.TODO));
    }

    /**
     * Helper function for bulk-importing settings from arrays of strings.
     *
     * @param imported An array of symbol, name, next symbol, status type
     */
    static createFromImportedValue(imported: StatusCollectionEntry) {
        const symbol = imported[0];
        const type = Status.getTypeFromStatusTypeString(imported[3]);
        return new Status(new StatusConfiguration(symbol, imported[1], imported[2], false, type));
    }

    /**
     * Returns the completion status for a task, this is only supported
     * when the task is done/x.
     *
     * @return {*}  {boolean}
     * @memberof Status
     */
    public isCompleted(): boolean {
        return this.type === StatusType.DONE;
    }

    /**
     * Compare all the fields in another Status, to detect any differences from this one.
     *
     * If any field is different in any way, it will return false.
     *
     * @param other
     */
    public identicalTo(other: Status): boolean {
        return this.configuration.identicalTo(other.configuration);
    }

    /**
     * Return a one-line summary of the status, for presentation to users.
     */
    public previewText() {
        let commandNotice = '';
        if (Status.tasksPluginCanCreateCommandsForStatuses() && this.availableAsCommand) {
            commandNotice = ' Available as a command.';
        }
        return (
            `- [${this.symbol}]` + // comment to break line
            ` => [${this.nextStatusSymbol}],` +
            ` name: '${this.name}',` +
            ` type: '${this.configuration.type}'.` +
            `${commandNotice}`
        );
    }

    /**
     * Whether Tasks can yet create 'Toggle Status' commands for statuses
     *
     * This is not yet possible, and so some UI features are temporarily hidden.
     * See https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1486
     * Once that issue is addressed, this method can be removed.
     */
    public static tasksPluginCanCreateCommandsForStatuses(): boolean {
        return false;
    }
}
