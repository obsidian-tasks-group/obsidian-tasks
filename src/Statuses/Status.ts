import { StatusConfiguration, StatusType } from './StatusConfiguration';
import type { StatusCollectionEntry } from './StatusCollection';

/**
 * Tracks the possible states that a task can be in.
 *
 * Related classes:
 * @see StatusConfiguration
 * @see StatusRegistry
 * @see StatusSettings
 * @see StatusSettingsHelpers.ts
 * @see CustomStatusModal
 *
 * @class Status
 */
export class Status {
    /**
     * The default Done status. Goes to Todo when toggled.
     *
     * @static
     * @type {Status}
     */
    public static readonly DONE: Status = new Status(new StatusConfiguration('x', 'Done', ' ', true, StatusType.DONE));

    /**
     * A default status of empty, used when things go wrong.
     *
     * @static
     * @type {Status}
     */
    public static readonly EMPTY: Status = new Status(new StatusConfiguration('', 'EMPTY', '', true, StatusType.EMPTY));

    /**
     * The default Todo status. Goes to Done when toggled.
     * User may later be able to override this to go to In Progress instead.
     *
     * @static
     * @type {Status}
     */
    public static readonly TODO: Status = new Status(new StatusConfiguration(' ', 'Todo', 'x', true, StatusType.TODO));

    /**
     * The default Cancelled status. Goes to Todo when toggled.
     *
     * @static
     * @type {Status}
     */
    public static readonly CANCELLED: Status = new Status(
        new StatusConfiguration('-', 'Cancelled', ' ', true, StatusType.CANCELLED),
    );

    /**
     * The default In Progress status. Goes to Done when toggled.
     *
     * @static
     * @type {Status}
     */
    public static readonly IN_PROGRESS: Status = new Status(
        new StatusConfiguration('/', 'In Progress', 'x', true, StatusType.IN_PROGRESS),
    );

    /**
     * The default On Hold status. Goes to Todo when toggled.
     *
     * @static
     * @type {Status}
     */
    public static readonly ON_HOLD: Status = new Status(
        new StatusConfiguration('h', 'On Hold', ' ', true, StatusType.ON_HOLD),
    );

    /**
     * A sample Non-Task status. Goes to NON_TASK when toggled.
     *
     * @static
     * @type {Status}
     */
    public static readonly NON_TASK: Status = new Status(
        new StatusConfiguration('Q', 'Non-Task', 'A', true, StatusType.NON_TASK),
    );

    /**
     * The configuration stored in the data.json file.
     *
     * @type {StatusConfiguration}
     */
    public readonly configuration: StatusConfiguration;

    /**
     * The symbol used between the two square brackets in the markdown task.
     *
     * @type {string}
     */
    public get symbol(): string {
        return this.configuration.symbol;
    }

    /**
     * Returns the name of the status for display purposes.
     *
     * @type {string}
     */
    public get name(): string {
        return this.configuration.name;
    }

    /**
     * Returns the next status for a task when toggled.
     *
     * @type {string}
     * @see nextSymbol
     */
    public get nextStatusSymbol(): string {
        return this.configuration.nextStatusSymbol;
    }

    /**
     * Returns the next status for a task when toggled.
     * This is an alias for {@link nextStatusSymbol} which is provided for brevity in user scripts.
     *
     * @type {string}
     * @see nextStatusSymbol
     */
    public get nextSymbol(): string {
        return this.configuration.nextStatusSymbol;
    }

    /**
     * If true then it is registered as a command that the user can map to.
     *
     * @type {boolean}
     */
    public get availableAsCommand(): boolean {
        return this.configuration.availableAsCommand;
    }

    /**
     * Returns the status type. See {@link StatusType} for details.
     */
    public get type(): StatusType {
        return this.configuration.type;
    }

    /**
     * Returns the text to be used to represent the {@link StatusType} in group headings.
     *
     * The status types are in the same order as given by 'group by status.type'.
     * This is provided as a convenience for use in custom grouping.
     */
    public get typeGroupText(): string {
        const type = this.type;
        let prefix: string;
        // Add a numeric prefix to sort in to a meaningful order for users
        switch (type) {
            case StatusType.IN_PROGRESS:
                prefix = '1';
                break;
            case StatusType.TODO:
                prefix = '2';
                break;
            case StatusType.ON_HOLD:
                prefix = '3';
                break;
            case StatusType.DONE:
                prefix = '4';
                break;
            case StatusType.CANCELLED:
                prefix = '5';
                break;
            case StatusType.NON_TASK:
                prefix = '6';
                break;
            case StatusType.EMPTY:
                prefix = '7';
                break;
        }
        // Text inside the %%..%% comments is used to control the sorting in both sorting of tasks and naming of groups.
        // The comments are hidden by Obsidian when the headings are rendered.
        return `%%${prefix}%%${type}`;
    }

    /**
     * Creates an instance of Status. The registry will be added later in the case
     * of the default statuses.
     *
     * @param {StatusConfiguration} configuration
     */
    constructor(configuration: StatusConfiguration) {
        this.configuration = configuration;
    }

    /**
     * Return the StatusType to use for a symbol, if it is not in the StatusRegistry.
     * The core symbols are recognised.
     * Other symbols are treated as StatusType.TODO
     * @param symbol
     */
    static getTypeForUnknownSymbol(symbol: string): StatusType {
        switch (symbol) {
            case 'x':
            case 'X':
                return StatusType.DONE;
            case '/':
                return StatusType.IN_PROGRESS;
            case '-':
                return StatusType.CANCELLED;
            case '':
                return StatusType.EMPTY;
            case ' ':
            default:
                return StatusType.TODO;
        }
    }

    /**
     * Convert text that was saved from a StatusType value back to a StatusType.
     * Returns StatusType.TODO if the string is not valid.
     * @param statusTypeAsString
     */
    static getTypeFromStatusTypeString(statusTypeAsString: string): StatusType {
        return StatusType[statusTypeAsString as keyof typeof StatusType] || StatusType.TODO;
    }

    /**
     * Create a Status representing the given, unknown symbol.
     *
     * This can be useful when StatusRegistry does not recognise a symbol,
     * and we do not want to expose the user's data to the Status.EMPTY status.
     *
     * The type is set to TODO.
     * @param unknownSymbol
     */
    static createUnknownStatus(unknownSymbol: string) {
        return new Status(new StatusConfiguration(unknownSymbol, 'Unknown', 'x', false, StatusType.TODO));
    }

    /**
     * Helper function for bulk-importing settings from arrays of strings.
     *
     * @param imported An array of symbol, name, next symbol, status type
     */
    static createFromImportedValue(imported: StatusCollectionEntry) {
        const symbol = imported[0];
        const type = Status.getTypeFromStatusTypeString(imported[3]);
        return new Status(new StatusConfiguration(symbol, imported[1], imported[2], false, type));
    }

    /**
     * Returns the completion status for a task, this is only supported
     * when the task is done/x.
     *
     * @return {*}  {boolean}
     */
    public isCompleted(): boolean {
        return this.type === StatusType.DONE;
    }

    /**
     * Whether the task status type is {@link CANCELLED}.
     */
    public isCancelled(): boolean {
        return this.type === StatusType.CANCELLED;
    }

    /**
     * Compare all the fields in another Status, to detect any differences from this one.
     *
     * If any field is different in any way, it will return false.
     *
     * @param other
     */
    public identicalTo(other: Status): boolean {
        const args: Array<keyof StatusConfiguration> = [
            'symbol',
            'name',
            'nextStatusSymbol',
            'availableAsCommand',
            'type',
        ];
        for (const el of args) {
            if (this[el] !== other[el]) return false;
        }
        return true;
    }

    /**
     * Return a one-line summary of the status, for presentation to users.
     */
    public previewText() {
        let commandNotice = '';
        if (Status.tasksPluginCanCreateCommandsForStatuses() && this.availableAsCommand) {
            commandNotice = ' Available as a command.';
        }
        return (
            `- [${this.symbol}]` + // comment to break line
            ` => [${this.nextStatusSymbol}],` +
            ` name: '${this.name}',` +
            ` type: '${this.configuration.type}'.` +
            `${commandNotice}`
        );
    }

    /**
     * Whether Tasks can yet create 'Toggle Status' commands for statuses
     *
     * This is not yet possible, and so some UI features are temporarily hidden.
     * See https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1486
     * Once that issue is addressed, this method can be removed.
     */
    public static tasksPluginCanCreateCommandsForStatuses(): boolean {
        return false;
    }
}
