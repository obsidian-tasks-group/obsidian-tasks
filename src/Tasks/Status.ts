export enum Status {
    TODO = 'TODO',
    DONE = 'DONE',
}
