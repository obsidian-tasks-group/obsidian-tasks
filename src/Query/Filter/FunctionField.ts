import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { Grouper } from '../Grouper';
import { Explanation } from '../Explain/Explanation';
import { TaskExpression, parseAndEvaluateExpression } from '../../Scripting/TaskExpression';
import type { QueryContext } from '../../Scripting/QueryContext';
import type { SearchInfo } from '../SearchInfo';
import { Field } from './Field';
import { Filter, type FilterFunction } from './Filter';
import { FilterOrErrorMessage } from './FilterOrErrorMessage';

/**
 * A {@link Field} implement that accepts a JavaScript expression to filter or group tasks.
 *
 * See also {@link parseAndEvaluateExpression}
 */
export class FunctionField extends Field {
    createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        const match = Field.getMatch(this.filterRegExp(), line);
        if (match === null) {
            return FilterOrErrorMessage.fromError(line, 'Unable to parse line');
        }

        const expression = match[1];
        const taskExpression = new TaskExpression(expression);
        if (!taskExpression.isValid()) {
            return FilterOrErrorMessage.fromError(line, taskExpression.parseError!);
        }

        return FilterOrErrorMessage.fromFilter(
            new Filter(line, createFilterFunctionFromLine(taskExpression), new Explanation(line)),
        );
    }

    fieldName(): string {
        return 'function';
    }

    protected filterRegExp(): RegExp | null {
        return new RegExp(`^filter by ${this.fieldNameSingularEscaped()} (.*)`);
    }

    // -----------------------------------------------------------------------------------------------------------------
    // Grouping
    // -----------------------------------------------------------------------------------------------------------------

    public supportsGrouping(): boolean {
        return true;
    }

    public createGrouperFromLine(line: string): Grouper | null {
        const match = Field.getMatch(this.grouperRegExp(), line);
        if (match === null) {
            return null;
        }
        const reverse = !!match[1];
        const args = match[2];
        return new Grouper('function', createGrouperFunctionFromLine(args), reverse);
    }

    protected grouperRegExp(): RegExp {
        return new RegExp(`^group by ${this.fieldNameSingularEscaped()}( reverse)? (.*)`);
    }

    /**
     * This method does not work for 'group by function' as the user's instruction line
     * is required in order to create the {@link GrouperFunction}.
     *
     * So this class overrides {@link createGrouperFromLine} instead.
     * @throws Error
     */
    public grouper(): GrouperFunction {
        throw Error('grouper() function not valid for FunctionField. Use createGrouperFromLine() instead.');
    }
}

// -----------------------------------------------------------------------------------------------------------------
// Filtering
// -----------------------------------------------------------------------------------------------------------------

function createFilterFunctionFromLine(expression: TaskExpression): FilterFunction {
    return (task: Task, searchInfo: SearchInfo) => {
        const queryContext = searchInfo.queryContext();
        return filterByFunction(expression, task, queryContext);
    };
}

export function filterByFunction(expression: TaskExpression, task: Task, queryContext?: QueryContext): boolean {
    // Allow exceptions to propagate to caller, since this will be called in a tight loop.
    // In searches, it will be caught by Query.applyQueryToTasks().
    const result = expression.evaluate(task, queryContext);

    // We insist that 'filter by function' returns booleans,
    // to avoid users having to understand truthy and falsey values.
    if (typeof result === 'boolean') {
        return result;
    }

    throw Error(`filtering function must return true or false. This returned "${result}".`);
}

// -----------------------------------------------------------------------------------------------------------------
// Grouping
// -----------------------------------------------------------------------------------------------------------------

type GroupingArg = string;

function createGrouperFunctionFromLine(line: string): GrouperFunction {
    return (task: Task, searchInfo: SearchInfo) => {
        const queryContext = searchInfo.queryContext();
        return groupByFunction(task, line, queryContext);
    };
}

export function groupByFunction(task: Task, arg: GroupingArg, queryContext?: QueryContext): string[] {
    try {
        const result = parseAndEvaluateExpression(task, arg, queryContext);

        if (Array.isArray(result)) {
            return result.map((h) => h.toString());
        }

        // Task uses null to represent missing information.
        // So we treat null as an empty group or 'not in a heading', for simplicity for users.
        // This can be overridden with 'null || "No value"
        if (result === null) {
            return [];
        }

        // If there was an error in the expression, like it referred to
        // an unknown task field, result will be undefined, and the call
        // on undefined.toString() will give an exception and a useful error
        // message below. This is a feature: it gives users feedback on the problem
        // instruction line.
        const group = result.toString();
        return [group];
    } catch (e) {
        const errorMessage = `Error: Failed calculating expression "${arg}". The error message was: `;
        if (e instanceof Error) {
            return [errorMessage + e.message];
        } else {
            return [errorMessage + 'Unknown error'];
        }
    }
}
