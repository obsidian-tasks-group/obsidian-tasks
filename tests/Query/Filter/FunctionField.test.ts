/**
 * @jest-environment jsdom
 */
import moment from 'moment';

import { FunctionField } from '../../../src/Query/Filter/FunctionField';
import { Status } from '../../../src/Status';
import { Priority } from '../../../src/Task';
import { toGroupTaskFromBuilder, toGroupTaskWithPath } from '../../CustomMatchers/CustomMatchersForGrouping';
import { TaskBuilder } from '../../TestingTools/TaskBuilder';

window.moment = moment;

// -----------------------------------------------------------------------------------------------------------------
// Filtering
// -----------------------------------------------------------------------------------------------------------------

describe('FunctionField - filtering', () => {
    const functionField = new FunctionField();

    it('filter by function - with valid query', () => {
        const filter = functionField.createFilterOrErrorMessage('filter by function task.description.length > 5');
        expect(filter).toBeValid();
        expect(filter).toMatchTaskWithDescription('123456');
        expect(filter).not.toMatchTaskWithDescription('12345');
        expect(filter).not.toMatchTaskWithDescription('1234');
    });

    it('filter by function - should report syntax errors via FilterOrErrorMessage', () => {
        const instructionWithExtraClosingParen = 'filter by function task.status.name.toUpperCase())';
        const filter = functionField.createFilterOrErrorMessage(instructionWithExtraClosingParen);
        expect(filter).not.toBeValid();
        expect(filter.error).toEqual(
            'Error: Failed parsing expression "task.status.name.toUpperCase())".\nThe error message was:\n    "SyntaxError: Unexpected token \')\'"',
        );
    });

    it('filter by function - should report searching errors via exception', () => {
        // Arrange
        const instructionWithExtraClosingParen = 'filter by function task.wibble';
        const filter = functionField.createFilterOrErrorMessage(instructionWithExtraClosingParen);

        // Assert
        expect(filter).toBeValid();
        const t = () => {
            filter.filterFunction!(new TaskBuilder().build());
        };
        expect(t).toThrow(Error);
        expect(t).toThrowError('filtering function must return true or false. This returned "undefined".');
    });
});

// -----------------------------------------------------------------------------------------------------------------
// Sorting
// -----------------------------------------------------------------------------------------------------------------

describe('FunctionField - sorting', () => {
    it('should not support sorting', () => {
        const functionField = new FunctionField();
        expect(functionField.supportsSorting()).toEqual(false);
    });
});

// -----------------------------------------------------------------------------------------------------------------
// Grouping
// -----------------------------------------------------------------------------------------------------------------

afterEach(() => {
    jest.useRealTimers();
});

function createGrouper(line: string) {
    const grouper = new FunctionField().createGrouperFromLine(line);
    expect(grouper).not.toBeNull();
    return grouper;
}

describe('FunctionField - grouping - basics', () => {
    it('should support grouping', () => {
        const field = new FunctionField();
        expect(field.supportsGrouping()).toEqual(true);
    });

    it('should parse "group by function" line', () => {
        // Arrange
        const field = new FunctionField();
        const instruction = 'group by function task.path.startsWith("journal/") ? "journal/" : task.path';

        // Assert
        const grouper = field.createGrouperFromLine(instruction);
        expect(grouper).not.toBeNull();
        expect(grouper?.reverse).toEqual(false);
    });

    it('should parse "group by function reverse" line', () => {
        // Arrange
        const field = new FunctionField();
        const instruction = 'group by function reverse task.path.startsWith("journal/") ? "journal/" : task.path';

        // Assert
        const grouper = field.createGrouperFromLine(instruction);
        expect(grouper).not.toBeNull();
        expect(grouper?.reverse).toEqual(true);
    });
});

describe('FunctionField - grouping - error-handling', () => {
    it('should throw in grouper()', () => {
        // It's not possible to create a GrouperFunction without a line,
        // so grouper() method needs to throw and report it cannot work.
        const field = new FunctionField();
        const t = () => {
            field.grouper();
        };
        expect(t).toThrow(Error);
    });

    it('should give a meaningful error for invalid syntax', () => {
        const line = 'group by function hello';
        const grouper = createGrouper(line);
        toGroupTaskWithPath(grouper, 'journal/a/b.md', [
            'Error: Failed calculating expression "hello".\nThe error message was:\n    "ReferenceError: hello is not defined"',
        ]);
    });

    it('should give a meaningful error for unknown task field', () => {
        const line = 'group by function task.nonExistentField';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder(), [
            'Error: Failed calculating expression "task.nonExistentField". The error message was: Cannot read properties of undefined (reading \'toString\')',
        ]);
    });
});

describe('FunctionField - grouping return types', () => {
    const expressionsAndResults = [
        ['"hello"', ['hello']],
        ['""', ['']],
        ['[]', []],
        ['"" || "No value"', ['No value']],
        ['false', ['false']],
        ['true', ['true']],
        ['1', ['1']],
        ['0', ['0']],
        ['0 || "No value"', ['No value']],
        ['1.0765456', ['1.0765456']],
        ['1.0765456.toFixed(3)', ['1.077']],
        ['["heading1", "heading2"]', ['heading1', 'heading2']], // return two headings, indicating that this task should be displayed twice, once in each heading
        ['[1, 2]', ['1', '2']], // return two headings, that need to be converted to strings
        ['null', []],
        ['null || "No value"', ['No value']],
        [
            'undefined',
            [
                'Error: Failed calculating expression "undefined". The error message was: Cannot read properties of undefined (reading \'toString\')',
            ],
        ],
        ['undefined || "No value"', ['No value']],
        [
            // Should allow manual escaping of markdown
            String.raw`"I _am_ not _italic_".replaceAll("_", "\\_")`,
            [String.raw`I \_am\_ not \_italic\_`],
        ],
    ];

    it.each(expressionsAndResults)("expression: '%s'", (expression: any, result: any) => {
        const line = `group by function ${expression}`;
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder(), result);
    });
});

describe('FunctionField - grouping - example functions', () => {
    it('should display booleans in a meaningful way', () => {
        const line = "group by function task.status.symbol === '/'";
        const grouper = createGrouper(line);
        toGroupTaskWithPath(grouper, 'journal/a/b.md', ['false']);
    });

    it('using root and path', () => {
        // Note: This will change to task.file.path
        const line = 'group by function task.path.startsWith("journal/") ? "journal/" : task.path';
        const grouper = createGrouper(line);

        toGroupTaskWithPath(grouper, 'journal/a/b.md', ['journal/']);
        toGroupTaskWithPath(grouper, 'hello/world/from-me.md', ['hello/world/from-me.md']);
        toGroupTaskWithPath(grouper, 'file-in-root-folder.md', ['file-in-root-folder.md']);
    });

    it('using path stripping folder', () => {
        const line = 'group by function task.path.replace("some/prefix/", "")';
        const grouper = createGrouper(line);

        toGroupTaskWithPath(grouper, 'a/b/c.md', ['a/b/c.md']);
    });

    it('group by priority', () => {
        const line = 'group by function task.priority';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().priority(Priority.Highest), ['0']);
    });

    it('group by status symbol', () => {
        // A single space as the character in a heading is not useful, so replace with something displayable:
        const line = 'group by function task.status.symbol.replace(" ", "space")';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().status(Status.makeCancelled()), ['-']);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().status(Status.makeTodo()), ['space']);
    });

    it('group by status nextStatusSymbol', () => {
        const line = 'group by function task.status.nextStatusSymbol.replace(" ", "space")';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().status(Status.makeInProgress()), ['x']);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().status(Status.makeDone()), ['space']);
    });

    it('group by using number', () => {
        const line = 'group by function task.description.length';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().description('#task Hello'), ['11']);
    });

    it('group by using string literal', () => {
        const line = 'group by function reverse "Description length: " + task.description.length';
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().description('#task Hello'), ['Description length: 11']);
    });

    it('group by a selected tag', () => {
        const line = "group by function task.tags.find((tag) => tag.includes('#context/'))";
        const grouper = createGrouper(line);
        toGroupTaskFromBuilder(grouper, new TaskBuilder().tags(['#context/pc_home', '#topic/sys_admin']), [
            '#context/pc_home',
        ]);
    });
});
