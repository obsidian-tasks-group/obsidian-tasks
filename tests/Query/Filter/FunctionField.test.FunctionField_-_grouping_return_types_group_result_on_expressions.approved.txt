["hello",hello] => "hello",hello
["",] => "",
[[],] => [],
["" || "No value",No value] => "" || "No value",No value
[false,false] => false,false
[true,true] => true,true
[1,1] => 1,1
[0,0] => 0,0
[0 || "No value",No value] => 0 || "No value",No value
[1.0765456,1.0765456] => 1.0765456,1.0765456
[["heading1", "heading2"],heading1,heading2] => ["heading1", "heading2"],heading1,heading2
[[1, 2],1,2] => [1, 2],1,2
[null,Error: Failed calculating expression "null". The error message was: Cannot read properties of null (reading 'toString')] => null,Error: Failed calculating expression "null". The error message was: Cannot read properties of null (reading 'toString')
[null || "No value",No value] => null || "No value",No value
[undefined,Error: Failed calculating expression "undefined". The error message was: Cannot read properties of undefined (reading 'toString')] => undefined,Error: Failed calculating expression "undefined". The error message was: Cannot read properties of undefined (reading 'toString')
[undefined || "No value",No value] => undefined || "No value",No value
["I _am_ not _italic_".replaceAll("_", "\\_"),I \_am\_ not \_italic\_] => "I _am_ not _italic_".replaceAll("_", "\\_"),I \_am\_ not \_italic\_
